{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Before you turn this problem in, make sure everything runs as expected. First, **restart the kernel** (in the menubar, select Kernel$\\rightarrow$Restart) and then **run all cells** (in the menubar, select Cell$\\rightarrow$Run All).\n",
    "\n",
    "Make sure you fill in any place that says `YOUR CODE HERE` or \"YOUR ANSWER HERE\", as well as your name and collaborators below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "NAME = \"Abinesh Senthil Kumar\"\n",
    "COLLABORATORS = \"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "12b7e2cd4d9ae91a5a8db73dc8027335",
     "grade": false,
     "grade_id": "cell-871e03f0ab421082",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "# Lab 2 Introduction\n",
    "\n",
    "\n",
    "In this lab you will use parse a vcf file to extracts parts of it and load it into a database. \n",
    "The fields you will be parsing from the vcf file are: \n",
    "```\n",
    "CHROM\n",
    "POS\t\n",
    "ID\t\n",
    "REF\t\n",
    "ALT\t\n",
    "QUAL\t\n",
    "FILTER\n",
    "```\n",
    "and from the INFO column, the following fields:\n",
    "```\n",
    "1000g2015aug_all\n",
    "ExAC_ALL\n",
    "FATHMM_pred\n",
    "LRT_pred\n",
    "MetaLR_pred\n",
    "MetaSVM_pred\n",
    "MutationAssessor_pred\n",
    "MutationTaster_pred\n",
    "PROVEAN_pred\n",
    "Polyphen2_HDIV_pred\n",
    "Polyphen2_HVAR_pred\n",
    "SIFT_pred\n",
    "fathmm-MKL_coding_pred.\n",
    "```\n",
    "The fields:\n",
    "```\n",
    "FATHMM_pred\n",
    "LRT_pred\n",
    "MetaLR_pred\n",
    "MetaSVM_pred\n",
    "MutationAssessor_pred\n",
    "MutationTaster_pred\n",
    "PROVEAN_pred\n",
    "Polyphen2_HDIV_pred\n",
    "Polyphen2_HVAR_pred\n",
    "SIFT_pred\n",
    "fathmm-MKL_coding_pred\n",
    "```\n",
    "are predictor fields. They use a letter to indicate whether a given variation is harmful or not.\n",
    "\n",
    "NOTE: use the helper functions in cell2. To use them, you have to run CELL 2! \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "426d1dc6b36f5faa3f8057e38cfe1772",
     "grade": false,
     "grade_id": "cell-feb13f7a74dea8c4",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "## Helper functions\n",
    "\n",
    "import os\n",
    "import sqlite3\n",
    "from sqlite3 import Error\n",
    "import gzip\n",
    "\n",
    "def create_connection(db_file, delete_db=False):\n",
    "    if delete_db and os.path.exists(db_file):\n",
    "        os.remove(db_file)\n",
    "    conn = None\n",
    "    try:\n",
    "        conn = sqlite3.connect(db_file)\n",
    "        conn.execute(\"PRAGMA foreign_keys = 1\")\n",
    "    except Error as e:\n",
    "        print(e)\n",
    "\n",
    "    return conn\n",
    "\n",
    "\n",
    "def create_table(conn, create_table_sql):\n",
    "    try:\n",
    "        c = conn.cursor()\n",
    "        c.execute(create_table_sql)\n",
    "    except Error as e:\n",
    "        print(e)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "36cb99791ce23b9f9113fc6870c1cd39",
     "grade": false,
     "grade_id": "cell-61ba7bcac2d80e58",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "# Part 1 (10 pts)\n",
    "\n",
    "In part1 you will open the gzipped vcf file using the python gzip module and figure out all the possible values for the predictor fields. \n",
    "\n",
    "You do not have to unzip the file. Use `with gzip.open(filename,'rt') as fp:` to read one line at a time. \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "8177d929dc2e030a3e56658d1bc0a399",
     "grade": false,
     "grade_id": "cell-6220a32f79b3dbc2",
     "locked": false,
     "schema_version": 3,
     "solution": true,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "def get_predictor_values(filename):\n",
    "    \n",
    "    import gzip\n",
    "    with gzip.open(filename, 'rt') as fp:\n",
    "        file = list(fp)\n",
    "    ################\n",
    "        def format_sample_fields(format_field, sample_field):\n",
    "   \n",
    "            samplevalues = []\n",
    "            samplekey = []\n",
    "            for i,v in sample_field.items():\n",
    "                samplevalues.append(v)\n",
    "                samplekey.append(i)\n",
    "\n",
    "            format_field = format_field.split(\":\")\n",
    "\n",
    "\n",
    "            samplevalues = [i.split(\":\") for i in samplevalues]\n",
    "\n",
    "            c = []\n",
    "\n",
    "            for v in samplevalues:\n",
    "                c.append(v)\n",
    "\n",
    "            dict2 = []\n",
    "\n",
    "            for z in c:\n",
    "        \n",
    "                dict2.append(dict(zip(format_field , z)))\n",
    "    \n",
    "            finaldict = dict(zip(samplekey,dict2))\n",
    "    \n",
    "            return finaldict\n",
    "\n",
    "        #############\n",
    "        def create_dict_from_line(header, line):\n",
    "            keyas = []\n",
    "\n",
    "            for i in header:\n",
    "                if i == \"FORMAT\":\n",
    "                    break\n",
    "                keyas.append(i.strip())\n",
    "\n",
    "            keyas1 = []\n",
    "            for i in header:\n",
    "                keyas1.append(i.strip())\n",
    "    \n",
    "            keyas2 = keyas1[(len(keyas)+1):]\n",
    "\n",
    "            dictivaly = line.split(\"\\t\")\n",
    "\n",
    "            dictivaly1 = dictivaly[:len(keyas)]\n",
    "\n",
    "            dictionary = dict(zip(keyas,dictivaly1))\n",
    "\n",
    "            dictionary8th = str(dictivaly[len(keyas)])\n",
    "\n",
    "            dictionary2 = dict(zip(keyas2,dictivaly[(len(keyas)+1):]))\n",
    "\n",
    "            formatoutput1 = format_sample_fields(dictionary8th,dictionary2)\n",
    "\n",
    "            dict2nd = {'SAMPLE':formatoutput1}\n",
    "    \n",
    "            dictionary.update(dict2nd)\n",
    "    \n",
    "   \n",
    "            return dictionary\n",
    "        ############\n",
    "        def read_vcf_file(filename):\n",
    "    \n",
    "        #file = open(filename, 'r')\n",
    "    \n",
    "            listy = [i.strip() for i in file if i.strip()]\n",
    "            headery = [i.strip() for i in listy if i.startswith('#CHROM')]\n",
    "  \n",
    "            line1 = []    \n",
    "            for i in listy:\n",
    "                if i.startswith(\"##\"):\n",
    "                    continue\n",
    "                elif i.startswith('#CHROM'):\n",
    "                    continue\n",
    "                else:\n",
    "                    line1.append([i])\n",
    "            \n",
    "            finhead = (headery[0][1:]).split(\"\\t\")      \n",
    "            listdict = []\n",
    "            for linelist in line1:\n",
    "                for eachline in linelist:\n",
    "                    listdict.append(create_dict_from_line(finhead, eachline))\n",
    "            #print(type(listdict))        \n",
    "            return listdict\n",
    "    \n",
    "    data = read_vcf_file(file)\n",
    "    def extract_info_field(data):\n",
    "    \n",
    "        #print(data[1])\n",
    "    \n",
    "        infofield = []\n",
    "        for i in data:\n",
    "            infofield.append(i[\"INFO\"])\n",
    "        return infofield\n",
    "    ################\n",
    "    infofield = extract_info_field(data)\n",
    "    #print(infofield[0])\n",
    "    ########################\n",
    "    def create_dictionary_of_info_field_values(data):\n",
    "    \n",
    "        infofieldval = []\n",
    "        for i in data:\n",
    "            infofieldval.append(i.split(\";\"))\n",
    "    \n",
    "        for i in infofieldval:\n",
    "            count = 0\n",
    "            for y in i:\n",
    "            #print(type(y))\n",
    "                i[count] = y.split(\"=\")\n",
    "                count +=1 \n",
    "        \n",
    "    #print(infofieldval)\n",
    "        listnewdict = []\n",
    "        for i in infofieldval:\n",
    "            keys = []\n",
    "            values = []\n",
    "            for y in i:\n",
    "                try:\n",
    "                    values.append(y[1])\n",
    "                    keys.append(y[0])\n",
    "                \n",
    "                except:\n",
    "                    continue\n",
    "            newdict = (dict(zip(keys,values)))   \n",
    "            for key,value in newdict.items():\n",
    "                newdict[key] = value.replace(\"\\\\x3d\",\"=\")\n",
    "        \n",
    "            for key,value in newdict.items():\n",
    "                newdict[key] = value.replace(\"\\\\x3b\",\",\")\n",
    "    \n",
    "    \n",
    "            for key,value in newdict.items():\n",
    "                if newdict[key] == '.':\n",
    "                    newdict[key] = []\n",
    "      \n",
    "            for key,value in newdict.items():\n",
    "                if newdict[key] == []:\n",
    "                    continue;\n",
    "                else:\n",
    "                    newdict[key] = [value]  \n",
    "            listnewdict.append(newdict)\n",
    "    #print(newdict)\n",
    "        if len(data) == 1:\n",
    "            return newdict\n",
    "        else:\n",
    "            return listnewdict\n",
    "    #########################\n",
    "    ans = create_dictionary_of_info_field_values(infofield)\n",
    "   \n",
    "    \n",
    "    #########################\n",
    "    keys_save_value = [\n",
    "        'FATHMM_pred',\n",
    "        'LRT_pred',\n",
    "        'MetaLR_pred',\n",
    "        'MetaSVM_pred',\n",
    "        'MutationAssessor_pred',\n",
    "        'MutationTaster_pred',\n",
    "        'PROVEAN_pred',\n",
    "        'Polyphen2_HDIV_pred',\n",
    "        'Polyphen2_HVAR_pred',\n",
    "        'SIFT_pred',\n",
    "        'fathmm-MKL_coding_pred',\n",
    "    ]\n",
    "    dictodi = {i : [] for i in keys_save_value}\n",
    "    \n",
    "    for z in ans:\n",
    "        for i,j in z.items():\n",
    "            for q,r in dictodi.items():\n",
    "                if i == q:\n",
    "                    if len(j) != 0:\n",
    "                        if j[0] not in r:\n",
    "                        #print(len(j))\n",
    "                            r.append(j[0].strip())\n",
    "                        \n",
    "    \n",
    "        \n",
    "        \n",
    "    return dictodi\n",
    "            \n",
    "            \n",
    "    #print(p)\n",
    "    # YOUR CODE HERE\n",
    "    #raise NotImplementedError()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "516936ec4947c069cb1abe1becfd97ff",
     "grade": true,
     "grade_id": "cell-70269f3e534eaa6a",
     "locked": true,
     "points": 10,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "## Can take 30 seconds to run!\n",
    "expected_solution = {'SIFT_pred': ['D', 'T'], 'Polyphen2_HDIV_pred': ['D', 'B', 'P'], 'Polyphen2_HVAR_pred': ['D', 'B', 'P'], 'LRT_pred': ['D', 'N', 'U'], 'MutationTaster_pred': ['D', 'P', 'N', 'A'], 'MutationAssessor_pred': ['H', 'N', 'L', 'M'], 'FATHMM_pred': ['T', 'D'], 'PROVEAN_pred': ['D', 'N'], 'MetaSVM_pred': ['D', 'T'], 'MetaLR_pred': ['D', 'T'], 'fathmm-MKL_coding_pred': ['D', 'N']}\n",
    "filename = 'test_4families_annovar.vcf.gz'\n",
    "predictor_values = get_predictor_values(filename)\n",
    "assert predictor_values == expected_solution\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "02e49d29c39c13db3e1374b2d2d9c278",
     "grade": false,
     "grade_id": "cell-70ccec7c3ed8b326",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "# Part 2 (No points)\n",
    "\n",
    "You will now create 13 tables\n",
    "Tables 1-11 will be for :\n",
    "\n",
    "FATHMM_pred\n",
    "LRT_pred \n",
    "MetaLR_pred\n",
    "MetaSVM_pred\n",
    "MutationAssessor_pred\n",
    "MutationTaster_pred\n",
    "PROVEAN_pred\n",
    "Polyphen2_HDIV_pred\n",
    "Polyphen2_HVAR_pred\n",
    "SIFT_pred\n",
    "fathmm_MKL_coding_pred ## NOTE: I have replaced the dash with an underscore. \n",
    "\n",
    "The first column for each of these tables will be the name of the field + ID, e.g., FATHMM_predID. This\n",
    "column will be of type not null primary key. The second column will be called `prediction` and will be of\n",
    "type TEXT not null. The prediction values will be values you extracted for each of the fields in the previous step. \n",
    "For example, 'FATHMM_pred' has two prediction values 'T' and 'D'. Name the tables after their field name, e.g.,  \n",
    "call table that will contains values of FATHMM_pred `FATHMM_pred`.  Make sure to sort the values, meaning, D should \n",
    "be inserted before T. \n",
    "\n",
    "\n",
    "Table 12 will be the Variants table. \n",
    "The first column will be VariantID. \n",
    "\n",
    "The other columns will be:\n",
    "CHROM   \n",
    "POS \n",
    "ID  \n",
    "REF \n",
    "ALT \n",
    "QUAL    \n",
    "FILTER\n",
    "thousandg2015aug_all ##NOTE: a column name cannot start with a number, so you have to rename!\n",
    "ExAC_ALL\n",
    "\n",
    "\n",
    "Table 12 will have 11 more columns that relate to each of prediction table. Consider writing a utility function\n",
    "to fetch the primary key for a given prediction from each of the prediction table. Name each of the \n",
    "column should be the name of predictor + ID, e.g., FATHMM_predID. \n",
    "\n",
    "\n",
    "You have already deteremined their data type, so use that to set their data type. \n",
    "For integer use INTEGER. \n",
    "For float use REAL. \n",
    "For string use TEXT. \n",
    "\n",
    "Table 13  will be called PredictionStats. The first column will be PredictorStatsID INTEGER NOT NULL PRIMARY KEY. \n",
    "The second column will be VariantID. The third column will be PredictorName. The fourth column will be PredictorValue. \n",
    "This is not a normalized table!\n",
    "\n",
    "The prediction value will be a float value mapped from the prediction text.  Use the following information to \n",
    "assign values: \n",
    "REF: https://brb.nci.nih.gov/seqtools/colexpanno.html#dbnsfp\n",
    "\n",
    "FATHMM_pred\n",
    "- T = 0\n",
    "- D = 1\n",
    "\n",
    "LRT_pred \n",
    "- D = 1\n",
    "- N = 0\n",
    "- U = 0\n",
    "\n",
    "MetaLR_pred\n",
    "- T = 0\n",
    "- D = 1\n",
    "\n",
    "MetaSVM_pred\n",
    "- T = 0\n",
    "- D = 1\n",
    "\n",
    "MutationAssessor_pred\n",
    "- H = 1\n",
    "- N = 0\n",
    "- L = 0.25\n",
    "- M = 0.5\n",
    "\n",
    "MutationTaster_pred\n",
    "- D = 1\n",
    "- P = 0\n",
    "- N = 0\n",
    "- A = 1\n",
    "\n",
    "PROVEAN_pred\n",
    "- D = 1\n",
    "- N = 0\n",
    "\n",
    "Polyphen2_HDIV_pred\n",
    "- D = 1\n",
    "- B = 0\n",
    "- P = 0.5\n",
    "\n",
    "Polyphen2_HVAR_pred\n",
    "- D = 1\n",
    "- B = 0\n",
    "- P = 0.5\n",
    "\n",
    "SIFT_pred\n",
    "- D = 1\n",
    "- T = 0\n",
    "\n",
    "fathmm-MKL_coding_pred\n",
    "- D = 1\n",
    "- N = 0\n",
    "\n",
    "\n",
    "```\n",
    "\n",
    "prediction_mapping = {\n",
    "    'FATHMM_pred': {'T': 0, 'D': 1},\n",
    "    'MetaLR_pred': {'T': 0, 'D': 1},\n",
    "    'MetaSVM_pred': {'T': 0, 'D': 1},\n",
    "    'SIFT_pred': {'T': 0, 'D': 1},\n",
    "    'fathmm_MKL_coding_pred': {'D': 1, 'N': 0},\n",
    "    'LRT_pred': {'U': 0, 'N': 0, 'D': 1},\n",
    "    'MutationAssessor_pred': {'H': 1, 'N': 0, 'L': 0.25, 'M': 0.5},  \n",
    "    'MutationTaster_pred': {'D': 1, 'P': 0, 'N': 0, 'A': 1},\n",
    "    'PROVEAN_pred': {'D': 1, 'N': 0},\n",
    "    'Polyphen2_HDIV_pred': {'D': 1, 'B': 0, 'P': 0.5},\n",
    "    'Polyphen2_HVAR_pred': {'D': 1, 'B': 0, 'P': 0.5},\n",
    "}\n",
    "\n",
    "\n",
    "```\n",
    "\n",
    "\n",
    "The idea is that 11 predictors have been used to annotate all the variants in the file. This table combines all that information in one table. \n",
    "By grouping the table based on variantid and summing the prediction values mapped from the prediction text, you can find which variants have a consensus on their being deterimental. \n",
    "\n",
    "IMPORTANT: instead of fathmm-MKL_coding_pred use fathmm_MKL_coding_pred everywhere. \n",
    "HINT: You have to commit the changes you make to the table, otherwise your changes will not be saved. Consider wrapping all changes inside `with conn:`\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "b57dd1bbe1b388369abcae0c015ecb8b",
     "grade": false,
     "grade_id": "cell-61cb3b6701bce76d",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "# Creating lab2.db \n",
    "db_file = 'lab2.db'\n",
    "conn = create_connection(db_file, delete_db=True)\n",
    "conn.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "a0df480570e6750d16a514bd751d3a0f",
     "grade": false,
     "grade_id": "cell-d88abb3c867e6ef4",
     "locked": false,
     "schema_version": 3,
     "solution": true,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "def create_tables_1_11(db_file):\n",
    "    \"\"\"\n",
    "    Create tables 1 to 11.\n",
    "    \n",
    "    INPUT: db_file -- Name of database file -- eg. lab2.db\n",
    "    \n",
    "    \"\"\"\n",
    "    conn = sqlite3.connect(db_file)\n",
    "    cursor = conn.cursor()\n",
    "    cursor.execute(\"PRAGMA foreign_keys = 1\")\n",
    "    \n",
    "    with conn:\n",
    "        \n",
    "        create_table_fathmm_pred = \"\"\"CREATE TABLE FATHMM_pred(FATHMM_predID INTEGER NOT NULL PRIMARY KEY, PREDICTION TEXT NOT NULL);\n",
    "        INSERT INTO FATHMM_pred VALUES(1, 'D');\n",
    "        INSERT INTO FATHMM_pred VALUES(2, 'T');\"\"\"\n",
    "        \n",
    "        create_table_lrt_pred = \"\"\"CREATE TABLE LRT_pred(LRT_predID INTEGER NOT NULL PRIMARY KEY, PREDICTION TEXT NOT NULL);\n",
    "        INSERT INTO LRT_pred VALUES(1, 'D');\n",
    "        INSERT INTO LRT_pred VALUES(2, 'N');\n",
    "        INSERT INTO LRT_pred VALUES(3, 'U');\"\"\"\n",
    "        \n",
    "        create_table_metalr_pred = \"\"\"CREATE TABLE MetaLR_pred(MetaLR_predID INTEGER NOT NULL PRIMARY KEY, PREDICTION TEXT NOT NULL);\n",
    "        INSERT INTO MetaLR_pred VALUES(1, 'D');\n",
    "        INSERT INTO MetaLR_pred VALUES(2, 'T');       \n",
    "        \"\"\"\n",
    "        create_table_metasvm = \"\"\"CREATE TABLE MetaSVM_pred(MetaSVM_predID INTEGER NOT NULL PRIMARY KEY, PREDICTION TEXT NOT NULL);\n",
    "        INSERT INTO MetaSVM_pred VALUES(1, 'D');\n",
    "        INSERT INTO MetaSVM_pred VALUES(2, 'T');\n",
    "        \"\"\"\n",
    "        create_table_mutationassesor = \"\"\"CREATE TABLE MutationAssessor_pred(MutationAssessor_predID INTEGER NOT NULL PRIMARY KEY, PREDICTION TEXT NOT NULL);\n",
    "        INSERT INTO MutationAssessor_pred VALUES(1, 'H');\n",
    "        INSERT INTO MutationAssessor_pred VALUES(2, 'L');\n",
    "        INSERT INTO MutationAssessor_pred VALUES(3, 'M');\n",
    "        INSERT INTO MutationAssessor_pred VALUES(4, 'N');\n",
    "        \"\"\"\n",
    "        create_table_mutationtaster = \"\"\"CREATE TABLE MutationTaster_pred(MutationTaster_predID INTEGER NOT NULL PRIMARY KEY, PREDICTION TEXT NOT NULL);\n",
    "        INSERT INTO MutationTaster_pred VALUES(1, 'A');\n",
    "        INSERT INTO MutationTaster_pred VALUES(2, 'D');\n",
    "        INSERT INTO MutationTaster_pred VALUES(3, 'N');\n",
    "        INSERT INTO MutationTaster_pred VALUES(4, 'P');   \n",
    "        \"\"\"\n",
    "        create_table_provean = \"\"\"CREATE TABLE PROVEAN_pred(PROVEAN_predID INTEGER NOT NULL PRIMARY KEY, PREDICTION TEXT NOT NULL);\n",
    "        INSERT INTO PROVEAN_pred VALUES(1, 'D');\n",
    "        INSERT INTO PROVEAN_pred VALUES(2, 'N');      \n",
    "        \"\"\"\n",
    "        \n",
    "        create_table_polyphen = \"\"\"CREATE TABLE Polyphen2_HDIV_pred(Polyphen2_HDIV_predID INTEGER NOT NULL PRIMARY KEY, PREDICTION TEXT NOT NULL);\n",
    "        INSERT INTO Polyphen2_HDIV_pred VALUES(1, 'B');\n",
    "        INSERT INTO Polyphen2_HDIV_pred VALUES(2, 'D');\n",
    "        INSERT INTO Polyphen2_HDIV_pred VALUES(3, 'P');\n",
    "        \"\"\"\n",
    "    \n",
    "        create_table_polyphen_hvar = \"\"\"CREATE TABLE Polyphen2_HVAR_pred(Polyphen2_HVAR_predID INTEGER NOT NULL PRIMARY KEY, PREDICTION TEXT NOT NULL);\n",
    "        INSERT INTO Polyphen2_HVAR_pred VALUES(1, 'B');\n",
    "        INSERT INTO Polyphen2_HVAR_pred VALUES(2, 'D');\n",
    "        INSERT INTO Polyphen2_HVAR_pred VALUES(3, 'P');       \n",
    "        \"\"\"\n",
    "    \n",
    "        create_table_sift = \"\"\"CREATE TABLE SIFT_pred(SIFT_predID INTEGER NOT NULL PRIMARY KEY, PREDICTION TEXT NOT NULL);\n",
    "        INSERT INTO SIFT_pred VALUES(1, 'D');\n",
    "        INSERT INTO SIFT_pred VALUES(2, 'T');     \n",
    "        \"\"\"\n",
    "        \n",
    "        create_table_fathmmmkl = \"\"\"CREATE TABLE fathmm_MKL_coding_pred(fathmm_MKL_coding_predID INTEGER NOT NULL PRIMARY KEY, PREDICTION TEXT NOT NULL);\n",
    "        INSERT INTO fathmm_MKL_coding_pred VALUES(1, 'D');\n",
    "        INSERT INTO fathmm_MKL_coding_pred VALUES(2, 'N');\n",
    "        \"\"\"\n",
    "    \n",
    "        cursor.executescript(create_table_fathmm_pred)\n",
    "        cursor.executescript(create_table_lrt_pred)\n",
    "        cursor.executescript(create_table_metalr_pred)\n",
    "        cursor.executescript(create_table_metasvm)\n",
    "        cursor.executescript(create_table_mutationassesor)\n",
    "        cursor.executescript(create_table_mutationtaster)\n",
    "        cursor.executescript(create_table_provean)\n",
    "        cursor.executescript(create_table_polyphen)\n",
    "        cursor.executescript(create_table_polyphen_hvar)\n",
    "        cursor.executescript(create_table_sift)\n",
    "        cursor.executescript(create_table_fathmmmkl)\n",
    "    \n",
    "    \n",
    "    #cursor.executescript\n",
    "    #conn.commit()\n",
    "    #conn.close()\n",
    "\n",
    "    # YOUR CODE HERE\n",
    "    #raise NotImplementedError()\n",
    "\n",
    "db_file = 'lab2.db'    \n",
    "create_tables_1_11(db_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "aca29ed5ea6e60f075e133fec0052535",
     "grade": false,
     "grade_id": "cell-ad7a8e7ac0b01d55",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "# Part 3 (10 pts)\n",
    "\n",
    "Write a function that returns a dictionary that maps for a given predictor its letter prediction to foreign key value. \n",
    "Conver the dash in 'fathmm-MKL_coding_pred' to an underscore. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "3ea21bf0090fdaced225244d86e1594c",
     "grade": false,
     "grade_id": "cell-c1fed27f8a6c9e23",
     "locked": false,
     "schema_version": 3,
     "solution": true,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "def get_predictor_value_to_fk_map(db_file):\n",
    "    \"\"\"\n",
    "    See part 3 description \n",
    "    \n",
    "    INPUT: db_file -- Name of database file -- eg. lab2.db\n",
    "    \n",
    "    \"\"\"\n",
    "    conn = sqlite3.connect(db_file)\n",
    "    cursor = conn.cursor()\n",
    "    cursor.execute(\"SELECT name FROM sqlite_master where type = 'table'\")\n",
    "    tables = cursor.fetchall()\n",
    "    \n",
    "    listy = []\n",
    "    for i in range(len(tables)):\n",
    "        listy.append(tables[i][0])\n",
    "    \n",
    "    \n",
    "    dict1 = {}\n",
    "    for i in listy:\n",
    "        #print(name[0])\n",
    "        cursor.execute(f\"SELECT * FROM {i}\")\n",
    "        data = cursor.fetchall()\n",
    "        keys = []\n",
    "        values = []\n",
    "        for j in range(len(data)):\n",
    "            keys.append(data[j][1])\n",
    "            values.append(data[j][0])\n",
    "        dict1[i] = dict(zip(keys,values))\n",
    "        \n",
    "    \n",
    "    return dict1\n",
    "    #conn.commit()\n",
    "    #conn.close()\n",
    "    \n",
    "\n",
    "    \n",
    "        \n",
    "    \n",
    "    # YOUR CODE HERE\n",
    "    #raise NotImplementedError()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "17e4c380bec42efad7c85fbe769ae80d",
     "grade": true,
     "grade_id": "cell-d6d06a3e0545c576",
     "locked": true,
     "points": 10,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "expected_solution = {\n",
    "    'FATHMM_pred': {'D': 1, 'T': 2}, \n",
    "    'LRT_pred': {'D': 1, 'N': 2, 'U': 3}, \n",
    "    'MetaLR_pred': {'D': 1, 'T': 2}, \n",
    "    'MetaSVM_pred': {'D': 1, 'T': 2}, \n",
    "    'MutationAssessor_pred': {'H': 1, 'L': 2, 'M': 3, 'N': 4}, \n",
    "    'MutationTaster_pred': {'A': 1, 'D': 2, 'N': 3, 'P': 4}, \n",
    "    'PROVEAN_pred': {'D': 1, 'N': 2}, \n",
    "    'Polyphen2_HDIV_pred': {'B': 1, 'D': 2, 'P': 3}, \n",
    "    'Polyphen2_HVAR_pred': {'B': 1, 'D': 2, 'P': 3}, \n",
    "    'SIFT_pred': {'D': 1, 'T': 2}, \n",
    "    'fathmm_MKL_coding_pred': {'D': 1, 'N': 2}}\n",
    "\n",
    "db_file = 'lab2.db'\n",
    "predictor_fk_map = get_predictor_value_to_fk_map(db_file)\n",
    "assert predictor_fk_map == expected_solution\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "bd97e9f0266ecb95bfc6c421df17b062",
     "grade": false,
     "grade_id": "cell-15a658481488bcd7",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "# Part 4 (No Points)\n",
    "Create table 12 or the variants table. See description above\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "885c0214ddc955d8769e5b1f2014735d",
     "grade": false,
     "grade_id": "cell-86e648d05e51d138",
     "locked": false,
     "schema_version": 3,
     "solution": true,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "def create_variants_table(db_file):\n",
    "    \"\"\"\n",
    "    Part 4\n",
    "    \"\"\"\n",
    "    conn = sqlite3.connect(db_file)\n",
    "    cursor = conn.cursor()\n",
    "    cursor.execute(\"PRAGMA foreign_keys = 1\")\n",
    "    \n",
    "    \n",
    "    create_table_variants = \"\"\"CREATE TABLE Variants(\n",
    "    VariantID INTEGER NOT NULL PRIMARY KEY, \n",
    "    CHROM TEXT,\n",
    "    POS INTEGER,\n",
    "    ID  TEXT,\n",
    "    REF TEXT,\n",
    "    ALT TEXT,\n",
    "    QUAL REAL,\n",
    "    FILTER TEXT,\n",
    "    THOUSANDG2015AUG_ALL REAL, \n",
    "    ExAC_ALL REAL,\n",
    "    FATHMM_predID INTEGER,\n",
    "    LRT_predID INTEGER,\n",
    "    MetaLR_predID INTEGER,\n",
    "    MetaSVM_predID INTEGER,\n",
    "    MutationAssessor_predID INTEGER,\n",
    "    MutationTaster_predID INTEGER,\n",
    "    PROVEAN_predID INTEGER,\n",
    "    Polyphen2_HDIV_predID INTEGER,\n",
    "    Polyphen2_HVAR_predID INTEGER,\n",
    "    SIFT_predID INTEGER,\n",
    "    fathmm_MKL_coding_predID INTEGER)\n",
    "    \"\"\"\n",
    "    \n",
    "    with conn:\n",
    "        create_table(conn, create_table_variants)\n",
    "    \n",
    "    #conn.commit()\n",
    "    #conn.close()\n",
    "    # YOUR CODE HERE\n",
    "    #raise NotImplementedError()\n",
    "\n",
    "# create table\n",
    "db_file = 'lab2.db'\n",
    "create_variants_table(db_file)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "0a07c128ac410b03efbe7c79c8754ae8",
     "grade": false,
     "grade_id": "cell-72d2c7d909ec378f",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "# Part 5 (No Points)\n",
    "\n",
    "Create table 13 -- or the prediction stats table. See description above. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "c28e9241740b966cd49b76cbe4d26fca",
     "grade": false,
     "grade_id": "cell-9410edb2ed7c5bb3",
     "locked": false,
     "schema_version": 3,
     "solution": true,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "def create_predictionstats_table(db_file):\n",
    "    \"\"\"\n",
    "    Part 5   \n",
    "    \"\"\"\n",
    "    conn = sqlite3.connect(db_file)\n",
    "    cursor = conn.cursor()\n",
    "    cursor.execute(\"PRAGMA foreign_keys = 1\")\n",
    "    \n",
    "    create_table_predict = \"\"\"CREATE TABLE PredictionStats(\n",
    "    PredictorStatsID INTEGER NOT NULL PRIMARY KEY,\n",
    "    VariantID INTEGER,\n",
    "    PredictorName TEXT,\n",
    "    PredictorValue REAL);\"\"\"\n",
    "    \n",
    "    with conn:\n",
    "        create_table(conn, create_table_predict)\n",
    "    \n",
    "    #conn.commit()\n",
    "    #conn.close()\n",
    "\n",
    "    \n",
    "    \n",
    "    # YOUR CODE HERE\n",
    "    #raise NotImplementedError()\n",
    "    \n",
    "db_file = 'lab2.db'\n",
    "create_predictionstats_table(db_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "02ead5219e3bdfecc28282a49807354c",
     "grade": false,
     "grade_id": "cell-aadb54a613358152",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "# Part 6 (10 Points)\n",
    "\n",
    "Write a function to pull the following info fields given the whole info field. \n",
    "```\n",
    "values_to_pull = [\n",
    "        '1000g2015aug_all',\n",
    "        'ExAC_ALL',\n",
    "        'FATHMM_pred',\n",
    "        'LRT_pred',\n",
    "        'MetaLR_pred',\n",
    "        'MetaSVM_pred',\n",
    "        'MutationAssessor_pred',\n",
    "        'MutationTaster_pred',\n",
    "        'PROVEAN_pred',\n",
    "        'Polyphen2_HDIV_pred',\n",
    "        'Polyphen2_HVAR_pred',\n",
    "        'SIFT_pred',\n",
    "        'fathmm-MKL_coding_pred',\n",
    "    ]\n",
    "```\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "85c2cba7f4ac522754215f18c414e225",
     "grade": false,
     "grade_id": "cell-1a912b3aacdae96d",
     "locked": false,
     "schema_version": 3,
     "solution": true,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "sample_info_input = \"AC=2;AF=0.333;AN=6;BaseQRankSum=2.23;ClippingRankSum=0;DP=131;ExcessHet=3.9794;FS=2.831;MLEAC=2;MLEAF=0.333;MQ=60;MQRankSum=0;QD=12.06;ReadPosRankSum=-0.293;SOR=0.592;VQSLOD=21.79;culprit=MQ;DB;POSITIVE_TRAIN_SITE;ANNOVAR_DATE=2018-04-16;Func.refGene=exonic;Gene.refGene=MAST2;GeneDetail.refGene=.;ExonicFunc.refGene=nonsynonymous_SNV;AAChange.refGene=MAST2:NM_015112:exon29:c.G3910A:p.V1304M;Func.ensGene=exonic;Gene.ensGene=ENSG00000086015;GeneDetail.ensGene=.;ExonicFunc.ensGene=nonsynonymous_SNV;AAChange.ensGene=ENSG00000086015:ENST00000361297:exon29:c.G3910A:p.V1304M;cytoBand=1p34.1;gwasCatalog=.;tfbsConsSites=.;wgRna=.;targetScanS=.;Gene_symbol=.;OXPHOS_Complex=.;Ensembl_Gene_ID=.;Ensembl_Protein_ID=.;Uniprot_Name=.;Uniprot_ID=.;NCBI_Gene_ID=.;NCBI_Protein_ID=.;Gene_pos=.;AA_pos=.;AA_sub=.;Codon_sub=.;dbSNP_ID=.;PhyloP_46V=.;PhastCons_46V=.;PhyloP_100V=.;PhastCons_100V=.;SiteVar=.;PolyPhen2_prediction=.;PolyPhen2_score=.;SIFT_prediction=.;SIFT_score=.;FatHmm_prediction=.;FatHmm_score=.;PROVEAN_prediction=.;PROVEAN_score=.;MutAss_prediction=.;MutAss_score=.;EFIN_Swiss_Prot_Score=.;EFIN_Swiss_Prot_Prediction=.;EFIN_HumDiv_Score=.;EFIN_HumDiv_Prediction=.;CADD_score=.;CADD_Phred_score=.;CADD_prediction=.;Carol_prediction=.;Carol_score=.;Condel_score=.;Condel_pred=.;COVEC_WMV=.;COVEC_WMV_prediction=.;PolyPhen2_score_transf=.;PolyPhen2_pred_transf=.;SIFT_score_transf=.;SIFT_pred_transf=.;MutAss_score_transf=.;MutAss_pred_transf=.;Perc_coevo_Sites=.;Mean_MI_score=.;COSMIC_ID=.;Tumor_site=.;Examined_samples=.;Mutation_frequency=.;US=.;Status=.;Associated_disease=.;Presence_in_TD=.;Class_predicted=.;Prob_N=.;Prob_P=.;SIFT_score=0.034;SIFT_converted_rankscore=0.440;SIFT_pred=D;Polyphen2_HDIV_score=0.951;Polyphen2_HDIV_rankscore=0.520;Polyphen2_HDIV_pred=P;Polyphen2_HVAR_score=0.514;Polyphen2_HVAR_rankscore=0.462;Polyphen2_HVAR_pred=P;LRT_score=0.002;LRT_converted_rankscore=0.368;LRT_pred=N;MutationTaster_score=1.000;MutationTaster_converted_rankscore=0.810;MutationTaster_pred=D;MutationAssessor_score=1.67;MutationAssessor_score_rankscore=0.430;MutationAssessor_pred=L;FATHMM_score=1.36;FATHMM_converted_rankscore=0.344;FATHMM_pred=T;PROVEAN_score=-1.4;PROVEAN_converted_rankscore=0.346;PROVEAN_pred=N;VEST3_score=0.158;VEST3_rankscore=0.189;MetaSVM_score=-1.142;MetaSVM_rankscore=0.013;MetaSVM_pred=T;MetaLR_score=0.008;MetaLR_rankscore=0.029;MetaLR_pred=T;M-CAP_score=.;M-CAP_rankscore=.;M-CAP_pred=.;CADD_raw=4.716;CADD_raw_rankscore=0.632;CADD_phred=24.6;DANN_score=0.998;DANN_rankscore=0.927;fathmm-MKL_coding_score=0.900;fathmm-MKL_coding_rankscore=0.506;fathmm-MKL_coding_pred=D;Eigen_coding_or_noncoding=c;Eigen-raw=0.461;Eigen-PC-raw=0.469;GenoCanyon_score=1.000;GenoCanyon_score_rankscore=0.747;integrated_fitCons_score=0.672;integrated_fitCons_score_rankscore=0.522;integrated_confidence_value=0;GERP++_RS=4.22;GERP++_RS_rankscore=0.490;phyloP100way_vertebrate=4.989;phyloP100way_vertebrate_rankscore=0.634;phyloP20way_mammalian=1.047;phyloP20way_mammalian_rankscore=0.674;phastCons100way_vertebrate=1.000;phastCons100way_vertebrate_rankscore=0.715;phastCons20way_mammalian=0.999;phastCons20way_mammalian_rankscore=0.750;SiPhy_29way_logOdds=17.151;SiPhy_29way_logOdds_rankscore=0.866;Interpro_domain=.;GTEx_V6_gene=ENSG00000162415.6;GTEx_V6_tissue=Nerve_Tibial;esp6500siv2_all=0.0560;esp6500siv2_aa=0.0160;esp6500siv2_ea=0.0761;ExAC_ALL=0.0553;ExAC_AFR=0.0140;ExAC_AMR=0.0386;ExAC_EAS=0.0005;ExAC_FIN=0.0798;ExAC_NFE=0.0788;ExAC_OTH=0.0669;ExAC_SAS=0.0145;ExAC_nontcga_ALL=0.0541;ExAC_nontcga_AFR=0.0129;ExAC_nontcga_AMR=0.0379;ExAC_nontcga_EAS=0.0004;ExAC_nontcga_FIN=0.0798;ExAC_nontcga_NFE=0.0802;ExAC_nontcga_OTH=0.0716;ExAC_nontcga_SAS=0.0144;ExAC_nonpsych_ALL=0.0496;ExAC_nonpsych_AFR=0.0140;ExAC_nonpsych_AMR=0.0386;ExAC_nonpsych_EAS=0.0005;ExAC_nonpsych_FIN=0.0763;ExAC_nonpsych_NFE=0.0785;ExAC_nonpsych_OTH=0.0638;ExAC_nonpsych_SAS=0.0145;1000g2015aug_all=0.024361;1000g2015aug_afr=0.0038;1000g2015aug_amr=0.0461;1000g2015aug_eur=0.0795;1000g2015aug_sas=0.0041;CLNALLELEID=.;CLNDN=.;CLNDISDB=.;CLNREVSTAT=.;CLNSIG=.;dbscSNV_ADA_SCORE=.;dbscSNV_RF_SCORE=.;snp138NonFlagged=rs33931638;avsnp150=rs33931638;CADD13_RawScore=4.716301;CADD13_PHRED=24.6;Eigen=0.4614;REVEL=0.098;MCAP=.;Interpro_domain=.;ICGC_Id=.;ICGC_Occurrence=.;gnomAD_genome_ALL=0.0507;gnomAD_genome_AFR=0.0114;gnomAD_genome_AMR=0.0430;gnomAD_genome_ASJ=0.1159;gnomAD_genome_EAS=0;gnomAD_genome_FIN=0.0802;gnomAD_genome_NFE=0.0702;gnomAD_genome_OTH=0.0695;gerp++gt2=4.22;cosmic70=.;InterVar_automated=Benign;PVS1=0;PS1=0;PS2=0;PS3=0;PS4=0;PM1=0;PM2=0;PM3=0;PM4=0;PM5=0;PM6=0;PP1=0;PP2=0;PP3=0;PP4=0;PP5=0;BA1=1;BS1=1;BS2=0;BS3=0;BS4=0;BP1=0;BP2=0;BP3=0;BP4=0;BP5=0;BP6=0;BP7=0;Kaviar_AF=0.0552127;Kaviar_AC=8536;Kaviar_AN=154602;ALLELE_END\"\n",
    "values_to_pull = [\n",
    "        '1000g2015aug_all',\n",
    "        'ExAC_ALL',\n",
    "        'FATHMM_pred',\n",
    "        'LRT_pred',\n",
    "        'MetaLR_pred',\n",
    "        'MetaSVM_pred',\n",
    "        'MutationAssessor_pred',\n",
    "        'MutationTaster_pred',\n",
    "        'PROVEAN_pred',\n",
    "        'Polyphen2_HDIV_pred',\n",
    "        'Polyphen2_HVAR_pred',\n",
    "        'SIFT_pred',\n",
    "        'fathmm-MKL_coding_pred',\n",
    "    ]\n",
    "def pull_info_values(info):\n",
    "    \"\"\"\n",
    "    See part 6 description\n",
    "    \"\"\"\n",
    "    sample = info.split(\";\")\n",
    "    \n",
    "    sampley = [i.split(\"=\") for i in sample]\n",
    "        \n",
    "    list1 = []\n",
    "    list2 = []\n",
    "    for i in sampley:\n",
    "        for j in values_to_pull:\n",
    "            if i[0] == j:\n",
    "                if i[0] == '1000g2015aug_all':\n",
    "                    list1.append('thousandg2015aug_all')\n",
    "                    list2.append(i[1])\n",
    "                    \n",
    "                if i[0] == 'fathmm-MKL_coding_pred':\n",
    "                    list1.append('fathmm_MKL_coding_pred')\n",
    "                    list2.append(i[1])\n",
    "                    \n",
    "                else:\n",
    "                    if i[0] != '1000g2015aug_all':\n",
    "                        if i[0] != 'fathmm-MKL_coding_pred':\n",
    "                            list1.append(i[0].strip())\n",
    "                            list2.append(i[1].strip())\n",
    "                \n",
    "    findict = dict(zip(list1,list2))\n",
    "    #print(findict)\n",
    "    return findict\n",
    "    # YOUR CODE HERE\n",
    "    #raise NotImplementedError()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "2fe3587c07040e2cff40d0d2a5969ed5",
     "grade": true,
     "grade_id": "cell-daa4d9dd1bb8e921",
     "locked": true,
     "points": 10,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "sample_info_input = \"AC=2;AF=0.333;AN=6;BaseQRankSum=2.23;ClippingRankSum=0;DP=131;ExcessHet=3.9794;FS=2.831;MLEAC=2;MLEAF=0.333;MQ=60;MQRankSum=0;QD=12.06;ReadPosRankSum=-0.293;SOR=0.592;VQSLOD=21.79;culprit=MQ;DB;POSITIVE_TRAIN_SITE;ANNOVAR_DATE=2018-04-16;Func.refGene=exonic;Gene.refGene=MAST2;GeneDetail.refGene=.;ExonicFunc.refGene=nonsynonymous_SNV;AAChange.refGene=MAST2:NM_015112:exon29:c.G3910A:p.V1304M;Func.ensGene=exonic;Gene.ensGene=ENSG00000086015;GeneDetail.ensGene=.;ExonicFunc.ensGene=nonsynonymous_SNV;AAChange.ensGene=ENSG00000086015:ENST00000361297:exon29:c.G3910A:p.V1304M;cytoBand=1p34.1;gwasCatalog=.;tfbsConsSites=.;wgRna=.;targetScanS=.;Gene_symbol=.;OXPHOS_Complex=.;Ensembl_Gene_ID=.;Ensembl_Protein_ID=.;Uniprot_Name=.;Uniprot_ID=.;NCBI_Gene_ID=.;NCBI_Protein_ID=.;Gene_pos=.;AA_pos=.;AA_sub=.;Codon_sub=.;dbSNP_ID=.;PhyloP_46V=.;PhastCons_46V=.;PhyloP_100V=.;PhastCons_100V=.;SiteVar=.;PolyPhen2_prediction=.;PolyPhen2_score=.;SIFT_prediction=.;SIFT_score=.;FatHmm_prediction=.;FatHmm_score=.;PROVEAN_prediction=.;PROVEAN_score=.;MutAss_prediction=.;MutAss_score=.;EFIN_Swiss_Prot_Score=.;EFIN_Swiss_Prot_Prediction=.;EFIN_HumDiv_Score=.;EFIN_HumDiv_Prediction=.;CADD_score=.;CADD_Phred_score=.;CADD_prediction=.;Carol_prediction=.;Carol_score=.;Condel_score=.;Condel_pred=.;COVEC_WMV=.;COVEC_WMV_prediction=.;PolyPhen2_score_transf=.;PolyPhen2_pred_transf=.;SIFT_score_transf=.;SIFT_pred_transf=.;MutAss_score_transf=.;MutAss_pred_transf=.;Perc_coevo_Sites=.;Mean_MI_score=.;COSMIC_ID=.;Tumor_site=.;Examined_samples=.;Mutation_frequency=.;US=.;Status=.;Associated_disease=.;Presence_in_TD=.;Class_predicted=.;Prob_N=.;Prob_P=.;SIFT_score=0.034;SIFT_converted_rankscore=0.440;SIFT_pred=D;Polyphen2_HDIV_score=0.951;Polyphen2_HDIV_rankscore=0.520;Polyphen2_HDIV_pred=P;Polyphen2_HVAR_score=0.514;Polyphen2_HVAR_rankscore=0.462;Polyphen2_HVAR_pred=P;LRT_score=0.002;LRT_converted_rankscore=0.368;LRT_pred=N;MutationTaster_score=1.000;MutationTaster_converted_rankscore=0.810;MutationTaster_pred=D;MutationAssessor_score=1.67;MutationAssessor_score_rankscore=0.430;MutationAssessor_pred=L;FATHMM_score=1.36;FATHMM_converted_rankscore=0.344;FATHMM_pred=T;PROVEAN_score=-1.4;PROVEAN_converted_rankscore=0.346;PROVEAN_pred=N;VEST3_score=0.158;VEST3_rankscore=0.189;MetaSVM_score=-1.142;MetaSVM_rankscore=0.013;MetaSVM_pred=T;MetaLR_score=0.008;MetaLR_rankscore=0.029;MetaLR_pred=T;M-CAP_score=.;M-CAP_rankscore=.;M-CAP_pred=.;CADD_raw=4.716;CADD_raw_rankscore=0.632;CADD_phred=24.6;DANN_score=0.998;DANN_rankscore=0.927;fathmm-MKL_coding_score=0.900;fathmm-MKL_coding_rankscore=0.506;fathmm-MKL_coding_pred=D;Eigen_coding_or_noncoding=c;Eigen-raw=0.461;Eigen-PC-raw=0.469;GenoCanyon_score=1.000;GenoCanyon_score_rankscore=0.747;integrated_fitCons_score=0.672;integrated_fitCons_score_rankscore=0.522;integrated_confidence_value=0;GERP++_RS=4.22;GERP++_RS_rankscore=0.490;phyloP100way_vertebrate=4.989;phyloP100way_vertebrate_rankscore=0.634;phyloP20way_mammalian=1.047;phyloP20way_mammalian_rankscore=0.674;phastCons100way_vertebrate=1.000;phastCons100way_vertebrate_rankscore=0.715;phastCons20way_mammalian=0.999;phastCons20way_mammalian_rankscore=0.750;SiPhy_29way_logOdds=17.151;SiPhy_29way_logOdds_rankscore=0.866;Interpro_domain=.;GTEx_V6_gene=ENSG00000162415.6;GTEx_V6_tissue=Nerve_Tibial;esp6500siv2_all=0.0560;esp6500siv2_aa=0.0160;esp6500siv2_ea=0.0761;ExAC_ALL=0.0553;ExAC_AFR=0.0140;ExAC_AMR=0.0386;ExAC_EAS=0.0005;ExAC_FIN=0.0798;ExAC_NFE=0.0788;ExAC_OTH=0.0669;ExAC_SAS=0.0145;ExAC_nontcga_ALL=0.0541;ExAC_nontcga_AFR=0.0129;ExAC_nontcga_AMR=0.0379;ExAC_nontcga_EAS=0.0004;ExAC_nontcga_FIN=0.0798;ExAC_nontcga_NFE=0.0802;ExAC_nontcga_OTH=0.0716;ExAC_nontcga_SAS=0.0144;ExAC_nonpsych_ALL=0.0496;ExAC_nonpsych_AFR=0.0140;ExAC_nonpsych_AMR=0.0386;ExAC_nonpsych_EAS=0.0005;ExAC_nonpsych_FIN=0.0763;ExAC_nonpsych_NFE=0.0785;ExAC_nonpsych_OTH=0.0638;ExAC_nonpsych_SAS=0.0145;1000g2015aug_all=0.024361;1000g2015aug_afr=0.0038;1000g2015aug_amr=0.0461;1000g2015aug_eur=0.0795;1000g2015aug_sas=0.0041;CLNALLELEID=.;CLNDN=.;CLNDISDB=.;CLNREVSTAT=.;CLNSIG=.;dbscSNV_ADA_SCORE=.;dbscSNV_RF_SCORE=.;snp138NonFlagged=rs33931638;avsnp150=rs33931638;CADD13_RawScore=4.716301;CADD13_PHRED=24.6;Eigen=0.4614;REVEL=0.098;MCAP=.;Interpro_domain=.;ICGC_Id=.;ICGC_Occurrence=.;gnomAD_genome_ALL=0.0507;gnomAD_genome_AFR=0.0114;gnomAD_genome_AMR=0.0430;gnomAD_genome_ASJ=0.1159;gnomAD_genome_EAS=0;gnomAD_genome_FIN=0.0802;gnomAD_genome_NFE=0.0702;gnomAD_genome_OTH=0.0695;gerp++gt2=4.22;cosmic70=.;InterVar_automated=Benign;PVS1=0;PS1=0;PS2=0;PS3=0;PS4=0;PM1=0;PM2=0;PM3=0;PM4=0;PM5=0;PM6=0;PP1=0;PP2=0;PP3=0;PP4=0;PP5=0;BA1=1;BS1=1;BS2=0;BS3=0;BS4=0;BP1=0;BP2=0;BP3=0;BP4=0;BP5=0;BP6=0;BP7=0;Kaviar_AF=0.0552127;Kaviar_AC=8536;Kaviar_AN=154602;ALLELE_END\"\n",
    "\n",
    "expected_solution = {\n",
    "    'thousandg2015aug_all': '0.024361', \n",
    "    'ExAC_ALL': '0.0553', \n",
    "    'SIFT_pred': 'D', \n",
    "    'Polyphen2_HDIV_pred': 'P', \n",
    "    'Polyphen2_HVAR_pred': 'P', \n",
    "    'LRT_pred': 'N',\n",
    "    'MutationTaster_pred': 'D', \n",
    "    'MutationAssessor_pred': 'L', \n",
    "    'FATHMM_pred': 'T', \n",
    "    'PROVEAN_pred': 'N', \n",
    "    'MetaSVM_pred': 'T', \n",
    "    'MetaLR_pred': 'T', \n",
    "    'fathmm_MKL_coding_pred': 'D'\n",
    "}\n",
    "\n",
    "\n",
    "solution  = pull_info_values(sample_info_input)\n",
    "assert solution == expected_solution\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "7db3c8a9dd9b9bcad14056f34bfb1357",
     "grade": false,
     "grade_id": "cell-5c6808c0fa04ea2e",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "# Part 7 (10 points)\n",
    "\n",
    "Remember that to insert a record in SQLite, you have to use `cur.execute(sql, values)`, where `sql` is the insert statement and `values` is a list/tuple of values that will be substituted into the `sql` string wherever there is a question mark. \n",
    "\n",
    "Write a function that takes in as input: CHROM, POS, ID, REF, ALT, QUAL, FILTER, info_values and returns a list with the values in the following order:\n",
    "```\n",
    "CHROM \n",
    "POS\n",
    "ID\n",
    "REF\n",
    "ALT\n",
    "QUAL\n",
    "FILTER\n",
    "thousandg2015aug_all # 1000 has been replaced by the text thousand\n",
    "ExAC_ALL\n",
    "FATHMM_pred\n",
    "LRT_pred\n",
    "MetaLR_pred\n",
    "MetaSVM_pred\n",
    "MutationAssessor_pred\n",
    "MutationTaster_pred\n",
    "PROVEAN_pred\n",
    "Polyphen2_HDIV_pred\n",
    "Polyphen2_HVAR_pred\n",
    "SIFT_pred\n",
    "fathmm_MKL_coding_pred # note that the dash has been replaced by underscore\n",
    "```\n",
    "\n",
    "The info_values dictionary contains the predictor values. Use `None` for any empty/missing value for info fields, thousandg2015aug_all, and ExAC_ALL. \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "36cd8d7cf36596717f5cc7647ef3accb",
     "grade": false,
     "grade_id": "cell-64622ed1cd1edf93",
     "locked": false,
     "schema_version": 3,
     "solution": true,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "def build_values_list(CHROM, POS, ID, REF, ALT, QUAL, FILTER, info_values):\n",
    "    \"\"\"\n",
    "    See part 7 description \n",
    "    \n",
    "    \"\"\"\n",
    "    dicti = predictor_fk_map\n",
    "    #print(dicti)\n",
    "    \n",
    "    listy = [CHROM,POS,ID,REF,ALT,QUAL,FILTER]\n",
    "\n",
    "    for k,v in info_values.items():\n",
    "        try:\n",
    "            float(v)\n",
    "        except:\n",
    "            if '.' in v:\n",
    "                info_values[k] = None\n",
    "\n",
    "\n",
    "    for k,v in info_values.items():\n",
    "        \n",
    "        if k == 'thousandg2015aug_all':\n",
    "            listy.append(info_values[k])\n",
    "        \n",
    "        elif k == 'ExAC_ALL':\n",
    "            listy.append(info_values[k])\n",
    "\n",
    "    if 'thousandg2015aug_all' not in info_values.keys(): \n",
    "        listy.append(None)\n",
    "\n",
    "    if 'ExAC_ALL' not in info_values.keys():\n",
    "        listy.append(None)\n",
    "\n",
    "    for k,v in dicti.items():\n",
    "        \n",
    "        if k in info_values.keys():\n",
    "            value = info_values[k]\n",
    "        \n",
    "            if value in v.keys():\n",
    "                listy.append(v[value])\n",
    "            else:\n",
    "                listy.append(value)\n",
    "    #print(listy)\n",
    "    return listy\n",
    "    # YOUR CODE HERE\n",
    "    #raise NotImplementedError()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "93470c7dfca66199ca14302923f9bc8c",
     "grade": true,
     "grade_id": "cell-c318c4d870f5531b",
     "locked": true,
     "points": 10,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "CHROM, POS, ID, REF, ALT, QUAL, FILTER = (7, 87837848, '.', 'C', 'A', 418.25, 'PASS') \n",
    "info_values = {'SIFT_pred': 'D', 'Polyphen2_HDIV_pred': 'D', 'Polyphen2_HVAR_pred': 'D', 'LRT_pred': 'D', 'MutationTaster_pred': 'D', 'MutationAssessor_pred': 'H', 'FATHMM_pred': 'T', 'PROVEAN_pred': 'D', 'MetaSVM_pred': 'D', 'MetaLR_pred': 'D', 'fathmm_MKL_coding_pred': 'D'}\n",
    "\n",
    "results = build_values_list(CHROM, POS, ID, REF, ALT, QUAL, FILTER, info_values)\n",
    "expected_results = [7, 87837848, '.', 'C', 'A', 418.25, 'PASS', None, None, 2, 1, 1, 1, 1, 2, 1, 2, 2, 1, 1]\n",
    "assert results == expected_results\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "58711bb5e6292a433b4e091e05877edd",
     "grade": false,
     "grade_id": "cell-95cecd8da8b89c54",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "# Part 8 (No Points)\n",
    "\n",
    "Create a function that takes the database `conn` and `values` from the `build_values_list` function to insert a variant record. \n",
    "\n",
    "IMPORTANT: Function should return the id of the row inserted, which will be VariantId\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "e226af966d70250ffe5720197deb152e",
     "grade": false,
     "grade_id": "cell-72a446c435bc6eb5",
     "locked": false,
     "schema_version": 3,
     "solution": true,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "def insert_variant(conn, values):\n",
    "    \"\"\"\n",
    "    See description Part 8\n",
    "    \"\"\"\n",
    "    conn= create_connection(conn)\n",
    "    cursor = conn.cursor()\n",
    "    \n",
    "    sqlstatement = \"\"\"INSERT INTO VARIANTS (CHROM, POS , ID , REF , ALT , QUAL , FILTER , THOUSANDG2015AUG_ALL , ExAC_ALL, FATHMM_predID, LRT_predID,  MetaLR_predID, MetaSVM_predID, MutationAssessor_predID, MutationTaster_predID, PROVEAN_predID, Polyphen2_HDIV_predID, Polyphen2_HVAR_predID, SIFT_predID, fathmm_MKL_coding_predID) VALUES(?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)\"\"\"\n",
    "    \n",
    "    \n",
    "    #conn.commit()\n",
    "    \n",
    "    with conn:\n",
    "        cursor.execute(sqlstatement, values)\n",
    "        #cursor.execute(\"select max(variantid) from variants\")\n",
    "        \n",
    "        #ans = cursor.fetchall()\n",
    "    #print(cursor.fetchall())\n",
    "    #conn.commit()\n",
    "    #conn.close()\n",
    "    \n",
    "    #print(ans[0][0])\n",
    "    #print(cursor.lastrowid)\n",
    "    return cursor.lastrowid\n",
    "    \n",
    "    # YOUR CODE HERE\n",
    "    #raise NotImplementedError()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "c7535737ef8cea0c4b306bece88f78f5",
     "grade": false,
     "grade_id": "cell-3c360e7adb54ac73",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "# Part 9 (No Points)\n",
    "\n",
    "Create a function that takes the database `conn` and `values` which is the tuple `(VariantId, PredictorName, PredictorValue)` Where `VariantId` will be the return value from the `insert_variant` function,  `PredictorName` is the name of the predictor, and `PredictorValue` is the mapping of the text value to a numeric value. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "d54b688e19c0ffecfbd0642fcf900b18",
     "grade": false,
     "grade_id": "cell-6d17ca28ee066a3b",
     "locked": false,
     "schema_version": 3,
     "solution": true,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "def insert_predictionstat(conn, values):\n",
    "    \"\"\"\n",
    "    See description in part 9\n",
    "    \"\"\"\n",
    "    conn= create_connection(conn)\n",
    "    \n",
    "    cursor = conn.cursor()\n",
    "    \n",
    "    sqlstatement = \"INSERT INTO PredictionStats (VariantID, PredictorName, PredictorValue) VALUES (?, ?, ?)\"\n",
    "    \n",
    "    with conn:\n",
    "        cursor.execute(sqlstatement, values)\n",
    "    \n",
    "    #conn.commit()\n",
    "    \n",
    "    #conn.close()\n",
    "\n",
    "    \n",
    "    # YOUR CODE HERE\n",
    "    #raise NotImplementedError()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "ff5474de7b650cbb71238dc0bdee24a7",
     "grade": false,
     "grade_id": "cell-beb6aaa454ebf4e9",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "# Part 10 (No Points)\n",
    "\n",
    "Write a function to insert records into both the variants and predictor_stats table. \n",
    "Hint:\n",
    "1) Open connection to database  \n",
    "2) Read file one line at a time using gzip read\n",
    "3) Extract CHROM, POS, ID, REF, ALT, QUAL, FILTER, INFO\n",
    "4) Use the pull_info_values function\n",
    "5) Use build_values_list \n",
    "6) Use insert_variant -- save variant_id\n",
    "7) Use insert_predictionstat -- insert each predictor at a time and remember to use `prediction_mapping` mapping. \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "```\n",
    "\n",
    "prediction_mapping = {\n",
    "    'FATHMM_pred': {'T': 0, 'D': 1},\n",
    "    'MetaLR_pred': {'T': 0, 'D': 1},\n",
    "    'MetaSVM_pred': {'T': 0, 'D': 1},\n",
    "    'SIFT_pred': {'T': 0, 'D': 1},\n",
    "    'fathmm_MKL_coding_pred': {'D': 1, 'N': 0},\n",
    "    'LRT_pred': {'U': 0, 'N': 0, 'D': 1},\n",
    "    'MutationAssessor_pred': {'H': 1, 'N': 0, 'L': 0.25, 'M': 0.5},  \n",
    "    'MutationTaster_pred': {'D': 1, 'P': 0, 'N': 0, 'A': 1},\n",
    "    'PROVEAN_pred': {'D': 1, 'N': 0},\n",
    "    'Polyphen2_HDIV_pred': {'D': 1, 'B': 0, 'P': 0.5},\n",
    "    'Polyphen2_HVAR_pred': {'D': 1, 'B': 0, 'P': 0.5},\n",
    "}\n",
    "\n",
    "\n",
    "```\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "7c47c65c4488c8d6f4c76b6e577386c0",
     "grade": false,
     "grade_id": "cell-09175a401cb7fde3",
     "locked": false,
     "schema_version": 3,
     "solution": true,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "prediction_mapping = {\n",
    "    'FATHMM_pred': {'T': 0, 'D': 1},\n",
    "    'MetaLR_pred': {'T': 0, 'D': 1},\n",
    "    'MetaSVM_pred': {'T': 0, 'D': 1},\n",
    "    'SIFT_pred': {'T': 0, 'D': 1},\n",
    "    'fathmm_MKL_coding_pred': {'D': 1, 'N': 0},\n",
    "    'LRT_pred': {'U': 0, 'N': 0, 'D': 1},\n",
    "    'MutationAssessor_pred': {'H': 1, 'N': 0, 'L': 0.25, 'M': 0.5},  \n",
    "    'MutationTaster_pred': {'D': 1, 'P': 0, 'N': 0, 'A': 1},\n",
    "    'PROVEAN_pred': {'D': 1, 'N': 0},\n",
    "    'Polyphen2_HDIV_pred': {'D': 1, 'B': 0, 'P': 0.5},\n",
    "    'Polyphen2_HVAR_pred': {'D': 1, 'B': 0, 'P': 0.5},\n",
    "}\n",
    "\n",
    "    #with gzip.open(filename, 'rt') as fp:\n",
    "    #file = list(fp)\n",
    "    \n",
    "\n",
    "\n",
    "def populate_variants_predictorstats_tables(db_file, filename):\n",
    "    \"\"\"\n",
    "    See description in part 10\n",
    "    \"\"\"\n",
    "    # YOUR CODE HERE\n",
    "    \n",
    "    listy = []\n",
    "    \n",
    "    with gzip.open(filename, 'rt') as fp:\n",
    "        \n",
    "        for i in fp:\n",
    "            if i.startswith('#'):\n",
    "                continue\n",
    "            else:\n",
    "                listy.append(i)\n",
    "\n",
    "    #print(len(listy))\n",
    "\n",
    "    for j in listy:\n",
    "        \n",
    "        CHROM, POS, ID, REF, ALT, QUAL, FILTER = j.split()[0:7]\n",
    "        \n",
    "        info = j.split()[7]\n",
    "        \n",
    "        info_values = pull_info_values(info)\n",
    "        \n",
    "        values = build_values_list(CHROM, POS, ID, REF, ALT, QUAL, FILTER, info_values)\n",
    "        \n",
    "        \n",
    "        conn = create_connection(db_file)\n",
    "        #print(conn)\n",
    "        with conn:\n",
    "            variantid = insert_variant(db_file, values)\n",
    "\n",
    "    \n",
    "            for key,value in info_values.items():\n",
    "            \n",
    "                if key == 'thousandg2015aug_all' or  key == 'ExAC_ALL':\n",
    "                    continue\n",
    "        \n",
    "                if key in prediction_mapping.keys():\n",
    "            \n",
    "                    dictodi = prediction_mapping[key]\n",
    "                \n",
    "                    if value in dictodi.keys():\n",
    "                    \n",
    "                        name = key\n",
    "                    \n",
    "                        valuez = dictodi[value]\n",
    "                    \n",
    "                        values = (variantid, name, valuez)\n",
    "                    \n",
    "                        insert_predictionstat(db_file, values)\n",
    "\n",
    "    \n",
    "    #raise NotImplementedError()\n",
    "try:\n",
    "    conn.close()\n",
    "except:\n",
    "    pass\n",
    "\n",
    "db_file = 'lab2.db'\n",
    "filename = 'test_4families_annovar.vcf.gz'\n",
    "populate_variants_predictorstats_tables(db_file, filename)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "080cfb7285cf1838e0d0fb2262e0e42e",
     "grade": false,
     "grade_id": "cell-07b168b75e9dc969",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "# Part 11 (10 Points)\n",
    "\n",
    "Write a function that returns the total number of variants\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "d7c6644f9602c3a52f61a91a16d149ad",
     "grade": false,
     "grade_id": "cell-1d18f5ef024c5d12",
     "locked": false,
     "schema_version": 3,
     "solution": true,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "def num_of_total_variants(conn):\n",
    "    # YOUR CODE HERE\n",
    "    cursor = conn.cursor()\n",
    "    cursor.execute(\"Select count(*) from variants\" )\n",
    "    answer = cursor.fetchall()\n",
    "    print((answer))\n",
    "    return answer[0][0]\n",
    "#     cursor.execute(\"Select count(*) from Variants\" )\n",
    "    raise NotImplementedError()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "e0620bed125f1503a864aa02388e1f82",
     "grade": true,
     "grade_id": "cell-880f6f16bf3fe3e3",
     "locked": true,
     "points": 10,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[(50001,)]\n"
     ]
    }
   ],
   "source": [
    "db_file = 'lab2.db'\n",
    "conn = create_connection(db_file)\n",
    "assert num_of_total_variants(conn) == 50001\n",
    "conn.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "869c5ab8d3a5f454ea093610ec290c8f",
     "grade": false,
     "grade_id": "cell-014a6a5fbd32d466",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "# Part 12 (10 Points)\n",
    "Write a function returns the total number of variant predictions -- the count of the predictiostats table. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "7429092cc88f97347b136821086f9799",
     "grade": false,
     "grade_id": "cell-375db6c511630ece",
     "locked": false,
     "schema_version": 3,
     "solution": true,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "def num_of_total_variant_predictions(conn):\n",
    "    \"\"\"\n",
    "    Part 12\n",
    "    \"\"\"\n",
    "    cursor = conn.cursor()\n",
    "    cursor.execute(\"Select count(*) from predictionstats\" )\n",
    "    answer = cursor.fetchall()\n",
    "    \n",
    "    return (answer[0][0])\n",
    "    # YOUR CODE HERE\n",
    "    #raise NotImplementedError()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "3fd4a49acab13980a3a88d7a38bc4460",
     "grade": true,
     "grade_id": "cell-4a9db4195a7202c1",
     "locked": true,
     "points": 10,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "db_file = 'lab2.db'\n",
    "conn = create_connection(db_file)\n",
    "assert num_of_total_variant_predictions(conn) == 1324\n",
    "conn.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "ccfbd666071479a3e22ab6ea2000fe53",
     "grade": false,
     "grade_id": "cell-7eb7e0e33e847c55",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "# Part 13 (10 Points)\n",
    "Return the total number of variant predictions that have value greater than zero. Number of values from the predictionstats table that are greater than 0. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "622fff446baff1967ab305668d243e41",
     "grade": false,
     "grade_id": "cell-d511bb5062a366a8",
     "locked": false,
     "schema_version": 3,
     "solution": true,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "def num_of_total_variant_predictions_with_value_gt_zero(conn):\n",
    "    \"\"\"\n",
    "    See part 13 description\n",
    "    \"\"\"\n",
    "    cursor = conn.cursor()\n",
    "    \n",
    "    cursor.execute(\"select count(predictorvalue) from predictionstats where predictorvalue > 0\")\n",
    "    \n",
    "    data = cursor.fetchall()\n",
    "    #print(data)\n",
    "    return data[0][0]\n",
    "    \n",
    "    # YOUR CODE HERE\n",
    "#     raise NotImplementedError()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "af597d6d881e8d170a19031097f15d28",
     "grade": true,
     "grade_id": "cell-9f84bca2161e571b",
     "locked": true,
     "points": 10,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "db_file = 'lab2.db'\n",
    "conn = create_connection(db_file)\n",
    "assert num_of_total_variant_predictions_with_value_gt_zero(conn) == 219\n",
    "conn.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "cd1ef4605927b7d3c5fce2b39e2c9073",
     "grade": false,
     "grade_id": "cell-075254108290e0a9",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "# Part 14 (10 Points)\n",
    "\n",
    "Write a function that given `CHROM, POS, ID, REF, ALT` returns a variant's info with the following columns (column order is important) :\n",
    "```\n",
    "Variants.CHROM,\n",
    "Variants.POS,\n",
    "Variants.ID,\n",
    "Variants.REF,\n",
    "Variants.ALT,\n",
    "Variants.QUAL,\n",
    "Variants.FILTER,\n",
    "Variants.thousandg2015aug_all,\n",
    "Variants.ExAC_ALL,\n",
    "FATHMM_pred.prediction,\n",
    "LRT_pred.prediction,\n",
    "MetaLR_pred.prediction,\n",
    "MetaSVM_pred.prediction,\n",
    "MutationAssessor_pred.prediction,\n",
    "MutationTaster_pred.prediction,\n",
    "PROVEAN_pred.prediction,\n",
    "Polyphen2_HDIV_pred.prediction,\n",
    "Polyphen2_HVAR_pred.prediction,\n",
    "SIFT_pred.prediction,\n",
    "fathmm_MKL_coding_pred.prediction,\n",
    "sum(PredictionStats.PredictorValue)\n",
    "```\n",
    "\n",
    "For the predictions, return the actual text value. And the last column is the sum of all the mapped prediction scores for a given variant!\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "343581499fff2c4c58e23631e8119a74",
     "grade": false,
     "grade_id": "cell-e0384800793c5ef7",
     "locked": false,
     "schema_version": 3,
     "solution": true,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "def fetch_variant(conn, CHROM, POS, ID, REF, ALT):\n",
    "    \"\"\"\n",
    "    See Part 14 description\n",
    "    \"\"\"\n",
    "    cursor = conn.cursor()\n",
    "    cursor.execute(\"PRAGMA foreign_keys = 1\")\n",
    "    values = (CHROM, POS, ID, REF, ALT)\n",
    "    \n",
    "    \n",
    "    try:\n",
    "        sql_statement = \"\"\"select V.CHROM,V.POS,V.ID,V.REF,V.ALT,V.QUAL,V.FILTER,V.ExAC_ALL,V.thousandg2015aug_all,A.prediction,\n",
    "        B.prediction,\n",
    "        C.prediction,\n",
    "        D.prediction,\n",
    "        E.prediction,\n",
    "        F.prediction,\n",
    "        G.prediction,\n",
    "        H.prediction,\n",
    "        I.prediction,\n",
    "        J.prediction,\n",
    "        K.prediction,\n",
    "        sum(P.PredictorValue)\n",
    "        from VARIANTS V \n",
    "        inner join FATHMM_pred A on V.FATHMM_predID=A.FATHMM_predID \n",
    "        inner join LRT_pred B on V.LRT_predID=B.LRT_predID\n",
    "        inner join MetaLR_pred C on V.MetaLR_predID=C.MetaLR_predID\n",
    "        inner join MetaSVM_pred D on V.MetaSVM_predID=D.MetaSVM_predID\n",
    "        inner join MutationAssessor_pred E on V.MutationAssessor_predID=E.MutationAssessor_predID\n",
    "        inner join MutationTaster_pred F on V.MutationTaster_predID=F.MutationTaster_predID\n",
    "        inner join PROVEAN_pred G on V.PROVEAN_predID=G.PROVEAN_predID\n",
    "        inner join Polyphen2_HDIV_pred H on V.Polyphen2_HDIV_predID=H.Polyphen2_HDIV_predID\n",
    "        inner join Polyphen2_HVAR_pred I on V.Polyphen2_HVAR_predID=I.Polyphen2_HVAR_predID\n",
    "        inner join SIFT_pred J on V.SIFT_predID=J.SIFT_predID\n",
    "        inner join fathmm_MKL_coding_pred K on V.fathmm_MKL_coding_predID=K.fathmm_MKL_coding_predID \n",
    "        inner join PredictionStats P on V.VariantID=P.VariantID\n",
    "        WHERE CHROM=? AND POS =? AND ID=? AND REF =? AND ALT =?\n",
    "        Group by V.CHROM,V.POS,V.ID,V.REF,V.ALT\"\"\"\n",
    "    \n",
    "        cursor.execute(sql_statement, values)\n",
    "        table = cursor.fetchall()    \n",
    "    #print(table)\n",
    "        return (table[0])\n",
    "    \n",
    "    except:\n",
    "        sql_statement = 'SELECT * FROM VARIANTS WHERE CHROM=? AND POS =? AND ID=? AND REF =? AND ALT =? '\n",
    "        \n",
    "        cursor.execute(sql_statement, values)\n",
    "        \n",
    "        data = cursor.fetchall()\n",
    "        \n",
    "        #print(data)\n",
    "        data = list(data[0])\n",
    "        \n",
    "        data.append(None)\n",
    "        \n",
    "        return tuple(data[1:])\n",
    "    \n",
    "    \n",
    "    # YOUR CODE HERE\n",
    "    #raise NotImplementedError()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "e0e5210cf6b040adebcedfbc67c6c12e",
     "grade": true,
     "grade_id": "cell-42ef13abc31ede02",
     "locked": true,
     "points": 10,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "db_file = 'lab2.db'\n",
    "conn = create_connection(db_file)\n",
    "assert fetch_variant(conn, '22', 25599849, 'rs17670506', 'G', 'A') == ('22', 25599849, 'rs17670506', 'G', 'A', 3124.91, 'PASS', 0.0251597, 0.0425, 'D', 'D', 'T', 'T', 'M', 'D', 'D', 'D', 'D', 'D', 'D', 8.5)\n",
    "conn.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "eb9ca12ea887f3a91c1fb05680fb12e9",
     "grade": true,
     "grade_id": "cell-5fd3bc522d861215",
     "locked": true,
     "points": 10,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "db_file = 'lab2.db'\n",
    "conn = create_connection(db_file)\n",
    "assert fetch_variant(conn, 'X', 2836184, 'rs73632976', 'C', 'T') == ('X', 2836184, 'rs73632976', 'C', 'T', 1892.12, 'PASS', None, 0.0427, 'D', 'U', 'D', 'T', 'M', 'P', 'D', 'P', 'P', 'D', 'D', 6.5)\n",
    "conn.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "999f2d60e36e7d4deae9ade0de137adb",
     "grade": true,
     "grade_id": "cell-e3f85c8b7beab623",
     "locked": true,
     "points": 10,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "db_file = 'lab2.db'\n",
    "conn = create_connection(db_file)\n",
    "assert fetch_variant(conn, '5', 155935708, 'rs45559835', 'G', 'A') == ('5', 155935708, 'rs45559835', 'G', 'A', 1577.12, 'PASS', 0.0189696, 0.0451, 'D', 'D', 'T', 'T', 'L', 'D', 'D', 'P', 'B', 'T', 'D', 5.75)\n",
    "conn.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "b6657c6306c075c0598f140f966d70d2",
     "grade": true,
     "grade_id": "cell-500f892df9ffe02f",
     "locked": true,
     "points": 10,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "db_file = 'lab2.db'\n",
    "conn = create_connection(db_file)\n",
    "assert fetch_variant(conn, '4', 123416186, '.', 'A', 'G') == ('4', 123416186, '.', 'A', 'G', 23.25, 'PASS', None, None, None, None, None, None, None, None, None, None, None, None, None, None)\n",
    "conn.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "5b1550ea1c3ea3437d1633839f2a18a4",
     "grade": false,
     "grade_id": "cell-bcd262e0085a72c9",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "# Part 15 (10 Points)\n",
    "Write a function that returns the variant with the highest predictor score sum. \n",
    "\n",
    "Return the variant info the following order:\n",
    "```\n",
    "Variants.CHROM,\n",
    "Variants.POS,\n",
    "Variants.ID,\n",
    "Variants.REF,\n",
    "Variants.ALT,\n",
    "Variants.QUAL,\n",
    "Variants.FILTER,\n",
    "Variants.thousandg2015aug_all,\n",
    "Variants.ExAC_ALL,\n",
    "FATHMM_pred.prediction,\n",
    "LRT_pred.prediction,\n",
    "MetaLR_pred.prediction,\n",
    "MetaSVM_pred.prediction,\n",
    "MutationAssessor_pred.prediction,\n",
    "MutationTaster_pred.prediction,\n",
    "PROVEAN_pred.prediction,\n",
    "Polyphen2_HDIV_pred.prediction,\n",
    "Polyphen2_HVAR_pred.prediction,\n",
    "SIFT_pred.prediction,\n",
    "fathmm_MKL_coding_pred.prediction,\n",
    "sum(PredictionStats.PredictorValue)\n",
    "        \n",
    "```\n",
    "\n",
    "Again, return the predictor text values and the last column is the sum of the prediction values. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "77b8b1a4255ee71cfed8e862b3192c1f",
     "grade": false,
     "grade_id": "cell-6b1c044f90162322",
     "locked": false,
     "schema_version": 3,
     "solution": true,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "def variant_with_highest_sum_of_predictor_value(conn):\n",
    "    \"\"\"\n",
    "    See part 15 description \n",
    "    \"\"\"\n",
    "    cursor = conn.cursor()\n",
    "\n",
    "    sql_statement = \"\"\"select CHROM,POS,ID,REF,ALT,QUAL,FILTER,thousandg2015aug_all,Exac_ALL, b.prediction,\n",
    "    c.prediction,d.prediction,e.prediction,f.prediction,g.prediction,\n",
    "    h.prediction,i.prediction,j.prediction,k.prediction,l.prediction, m.pred \n",
    "    from variants a inner join (select variantid, sum(predictorvalue) as pred \n",
    "    from PredictionStats group by 1 order by 2 desc limit 1 ) \n",
    "    m on a.variantid=m.variantid inner join fathmm_pred b on a.fathmm_predid=b.fathmm_predid\n",
    "    inner join LRT_pred c on a.LRT_predid=c.LRT_predid inner join MetaLR_pred d on a.MetaLR_predid=d.MetaLR_predid \n",
    "    inner join MetaSVM_pred e on a.MetaSVM_predid=e.MetaSVM_predid inner join MutationAssessor_pred f \n",
    "    on a.MutationAssessor_predid=f.MutationAssessor_predid inner join MutationTaster_pred g \n",
    "    on a.MutationTaster_predid=g.MutationTaster_predid inner join PROVEAN_pred h on a.PROVEAN_predid=h.PROVEAN_predid \n",
    "    inner join Polyphen2_HDIV_pred i on a.Polyphen2_HDIV_predid=i.Polyphen2_HDIV_predid inner join Polyphen2_HVAR_pred j \n",
    "    on a.Polyphen2_HVAR_predid = j.Polyphen2_HVAR_predid inner join SIFT_pred k on a.SIFT_predid=k.SIFT_predid inner join \n",
    "    fathmm_MKL_coding_pred l on a.fathmm_MKL_coding_predid = l.fathmm_MKL_coding_predid;\"\"\"\n",
    "    \n",
    "    cursor.execute(sql_statement)\n",
    "    \n",
    "    ans = cursor.fetchall()\n",
    "    return ans[0]\n",
    "    # YOUR CODE HERE\n",
    "    #raise NotImplementedError()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "ca4c10b920229a189a27ef253dc3679c",
     "grade": true,
     "grade_id": "cell-9c7365267a16b42f",
     "locked": true,
     "points": 10,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "db_file = 'lab2.db'\n",
    "conn = create_connection(db_file)\n",
    "assert variant_with_highest_sum_of_predictor_value(conn) == ('7', 87837848, '.', 'C', 'A', 418.25, 'PASS', None, None, 'T', 'D', 'D', 'D', 'H', 'D', 'D', 'D', 'D', 'D', 'D', 10.0)\n",
    "conn.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "celltoolbar": "Raw Cell Format",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
